<a href="https://colab.research.google.com/github/abusufianrobin/Dataset-Analysis-and-Prediction/blob/main/ML_Project_3_Credit_Card_Fraud_Detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Importing the Dependencies


In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [3]:
# loading the dataset to a Pandas DataFrame
credit_card_data = pd.read_csv('/content/creditcard.csv')

In [4]:
# first 5 rows of the dataset
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

In [5]:
credit_card_data.tail()

Time        V1        V2        V3        V4        V5        V6  \
23853  32953  1.296815 -0.518508  0.348800  0.129750 -0.845934 -0.283287   
23854  32954  1.295646 -0.699613 -1.129649 -2.505043  1.428275  3.010605   
23855  32954 -1.691394 -1.452403  1.671861 -1.769590  1.636843 -0.943701   
23856  32954  1.112786  0.062772  1.481419  2.922471 -0.905121  0.366357   
23857  32954 -0.264220  1.107046  0.985544  0.346904  1.121035  0.069275   

             V7        V8        V9  ...       V21       V22       V23  \
23853 -0.463165 -0.007344 -0.530097  ... -0.498652 -0.791893  0.058456   
23854 -0.892932  0.798386  1.309837  ...  0.071540  0.120545 -0.144172   
23855  0.193116 -0.712756 -0.987725  ...  0.047745  0.539167 -0.698010   
23856 -0.670663  0.256586  0.599399  ...  0.010125  0.346740 -0.117334   
23857  1.307409 -0.681733  0.346650  ... -0.159083       NaN       NaN   

            V24       V25       V26       V27       V28  Amount  Class  
23853  0.067753  0.367979  0.400899  0.005089  0.008846   14.00    0.0  
23854  1.066577  0.615628  0.123208  0.013875  0.018547   37.07    0.0  
23855 -0.470563  0.621828 -0.254508 -0.724865 -0.360165   53.80    0.0  
23856  0.416103  0.586102  0.216021  0.042710  0.024984    0.00    0.0  
23857       NaN       NaN       NaN       NaN       NaN     NaN    NaN  

[5 rows x 31 columns]

In [6]:
# dataset informations
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23858 entries, 0 to 23857
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Time    23858 non-null  int64  
 1   V1      23858 non-null  float64
 2   V2      23858 non-null  float64
 3   V3      23858 non-null  float64
 4   V4      23858 non-null  float64
 5   V5      23858 non-null  float64
 6   V6      23858 non-null  float64
 7   V7      23858 non-null  float64
 8   V8      23858 non-null  float64
 9   V9      23858 non-null  float64
 10  V10     23858 non-null  float64
 11  V11     23858 non-null  float64
 12  V12     23858 non-null  float64
 13  V13     23858 non-null  float64
 14  V14     23858 non-null  float64
 15  V15     23858 non-null  float64
 16  V16     23858 non-null  float64
 17  V17     23858 non-null  float64
 18  V18     23858 non-null  float64
 19  V19     23858 non-null  float64
 20  V20     23858 non-null  float64
 21  V21     23858 non-null  float64
 22

In [7]:
# checking the number of missing values in each column
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       1
V23       1
V24       1
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64

In [8]:
# distribution of legit transactions & fraudulent transactions
credit_card_data['Class'].value_counts()

Class
0.0    23769
1.0       88
Name: count, dtype: int64

This Dataset is highly unblanced

0 --> Normal Transaction

1 --> fraudulent transaction

In [9]:
# separating the data for analysis
legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data.Class == 1]

In [10]:
print(legit.shape)
print(fraud.shape)

(23769, 31)
(88, 31)


In [11]:
# statistical measures of the data
legit.Amount.describe()

count    23769.000000
mean        73.880199
std        212.541174
min          0.000000
25%          6.000000
50%         18.110000
75%         65.850000
max       7879.420000
Name: Amount, dtype: float64

In [12]:
fraud.Amount.describe()

count      88.000000
mean      100.010000
std       265.845031
min         0.000000
25%         1.000000
50%         1.000000
75%        99.990000
max      1809.680000
Name: Amount, dtype: float64

In [13]:
# compare the values for both transactions
credit_card_data.groupby('Class').mean()

Time        V1        V2         V3        V4        V5  \
Class                                                                   
0.0    18213.77782 -0.208135  0.175983   0.774952  0.226463 -0.166866   
1.0    17935.87500 -8.613716  6.376169 -12.221731  6.231847 -6.027247   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.092304 -0.101868  0.006906  0.512306  ...  0.038311 -0.043337   
1.0   -2.487080 -8.308784  4.351326 -2.987199  ...  0.714069  0.539387   

            V22       V23       V24       V25      V26       V27       V28  \
Class                                                                        
0.0   -0.136382 -0.037279  0.014613  0.127111  0.02694  0.010111  0.004578   
1.0   -0.381823 -0.350615 -0.252970  0.346695  0.17976  0.856336  0.100578   

           Amount  
Class              
0.0     73.880199  
1.0    100.010000  

[2 rows x 30 columns]

Under-Sampling

Build a sample dataset containing similar distribution of normal transactions and Fraudulent Transactions

Number of Fraudulent Transactions --> 492

In [14]:
legit_sample = legit.sample(n=492)

Concatenating two DataFrames

In [15]:
new_dataset = pd.concat([legit_sample, fraud], axis=0)

In [16]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
7025    9208 -0.739303  0.245052  1.369126 -1.993243 -0.629139 -1.585142   
15351  26741  1.035747 -0.602512  0.665093  0.153916 -0.784367  0.038242   
10224  15923  0.894001  0.107164  0.573334  2.533359 -0.023634  0.321280   
16266  27663 -2.036914  0.027628  0.560256 -2.064557  0.342404 -0.621989   
10380  16704 -2.882783  2.837182  0.477824  0.949180  0.669877  2.097775   

             V7        V8        V9  ...       V21       V22       V23  \
7025   1.336233 -0.444693  1.673562  ... -0.098651 -0.292136  0.334449   
15351 -0.434992 -0.039034  0.742953  ...  0.109909  0.255093 -0.261417   
10224 -0.000609  0.040327  0.475115  ... -0.357872 -1.280754 -0.019893   
16266  0.698806 -0.358027 -0.638706  ... -0.437379 -0.676164  0.310383   
10380 -2.114552 -4.803090  0.072079  ... -2.680976  0.474545  0.220589   

            V24       V25       V26       V27       V28  Amount  Class  
7025   0.862811 -0.174093 -0.392752  0.007159  0.135741  160.86    0.0  
15351 -0.318887  0.323952  1.142553 -0.051847  0.032000  135.74    0.0  
10224 -0.485853  0.126119 -0.355911 -0.065971  0.033080  154.27    0.0  
16266 -0.605736  0.269125 -0.649222 -0.125239  0.250457   34.00    0.0  
10380 -1.487607 -0.244368 -0.377102 -0.611701  0.044168    0.77    0.0  

[5 rows x 31 columns]

In [17]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
18773  29753  0.269614  3.549755 -5.810353  5.809370  1.538808 -2.269219   
18809  29785  0.923764  0.344048 -2.880004  1.721680 -3.019565 -0.639736   
20198  30852 -2.830984  0.885657  1.199930  2.861292  0.321669  0.289966   
23308  32686  0.287953  1.728735 -1.652173  3.813544 -1.090927 -0.984745   
23422  32745 -2.179135  0.020218 -2.182733  2.572046 -3.663733  0.081568   

             V7        V8        V9  ...       V21       V22       V23  \
18773 -0.824203  0.351070 -3.759059  ...  0.371121 -0.322290 -0.549856   
18809 -3.801325  1.299096  0.864065  ...  0.899931  1.481271  0.725266   
20198  1.767760 -2.451050  0.069736  ...  0.546589  0.334971  0.172106   
23308 -2.202318  0.555088 -2.033892  ...  0.262202 -0.633528  0.092891   
23422  0.268049  0.660437 -2.374027  ...  1.026421  0.299614  1.656800   

            V24       V25       V26       V27       V28  Amount  Class  
18773 -0.520629  1.378210  0.564714  0.553255  0.402400    0.68    1.0  
18809  0.176960 -1.815638 -0.536517  0.489035 -0.049729   30.30    1.0  
20198  0.623590 -0.527114 -0.079215 -2.532445  0.311177  104.81    1.0  
23308  0.187613  0.368708 -0.132474  0.576561  0.309843    0.00    1.0  
23422  0.328433  0.106457  0.691775  0.196779  0.241085  717.15    1.0  

[5 rows x 31 columns]

In [18]:
new_dataset['Class'].value_counts()

Class
0.0    492
1.0     88
Name: count, dtype: int64

In [19]:
new_dataset.groupby('Class').mean()

Time        V1        V2         V3        V4        V5  \
Class                                                                    
0.0    18977.989837 -0.137936  0.067342   0.797722  0.230431 -0.288904   
1.0    17935.875000 -8.613716  6.376169 -12.221731  6.231847 -6.027247   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.011574 -0.088852 -0.005140  0.519319  ...  0.048930 -0.002950   
1.0   -2.487080 -8.308784  4.351326 -2.987199  ...  0.714069  0.539387   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.152579 -0.056834  0.011153  0.140089  0.067011  0.000067  0.006379   
1.0   -0.381823 -0.350615 -0.252970  0.346695  0.179760  0.856336  0.100578   

           Amount  
Class              
0.0     91.855488  
1.0    100.010000  

[2 rows x 30 columns]

Splitting the data into Features & Targets

In [20]:
X = new_dataset.drop(columns='Class', axis=1)
Y = new_dataset['Class']

In [21]:
print(X)

        Time        V1        V2        V3        V4        V5        V6  \
7025    9208 -0.739303  0.245052  1.369126 -1.993243 -0.629139 -1.585142   
15351  26741  1.035747 -0.602512  0.665093  0.153916 -0.784367  0.038242   
10224  15923  0.894001  0.107164  0.573334  2.533359 -0.023634  0.321280   
16266  27663 -2.036914  0.027628  0.560256 -2.064557  0.342404 -0.621989   
10380  16704 -2.882783  2.837182  0.477824  0.949180  0.669877  2.097775   
...      ...       ...       ...       ...       ...       ...       ...   
18773  29753  0.269614  3.549755 -5.810353  5.809370  1.538808 -2.269219   
18809  29785  0.923764  0.344048 -2.880004  1.721680 -3.019565 -0.639736   
20198  30852 -2.830984  0.885657  1.199930  2.861292  0.321669  0.289966   
23308  32686  0.287953  1.728735 -1.652173  3.813544 -1.090927 -0.984745   
23422  32745 -2.179135  0.020218 -2.182733  2.572046 -3.663733  0.081568   

             V7        V8        V9  ...       V20       V21       V22  \
7025   1.3362

In [22]:
print(Y)

7025     0.0
15351    0.0
10224    0.0
16266    0.0
10380    0.0
        ... 
18773    1.0
18809    1.0
20198    1.0
23308    1.0
23422    1.0
Name: Class, Length: 580, dtype: float64


Split the data into Training data & Testing Data

In [23]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

In [24]:
print(X.shape, X_train.shape, X_test.shape)

(580, 30) (464, 30) (116, 30)


Model Training

Logistic Regression

In [25]:
model = LogisticRegression()

In [28]:
# training the Logistic Regression Model with Training Data
model.fit(X_train, Y_train)

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

# **Output: **
class sklearn.linear_model.LogisticRegression(penalty='l2', *, dual=False, tol=0.0001, C=1.0, fit_intercept=True, intercept_scaling=1, class_weight=None, random_state=None, solver='lbfgs', max_iter=100, multi_class='deprecated', verbose=0, warm_start=False, n_jobs=None, l1_ratio=None)

Model Evaluation



Accuracy Score

In [29]:
# accuracy on training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [30]:
print('Accuracy on Training data : ', training_data_accuracy)

Accuracy on Training data :  0.9913793103448276


In [31]:
# accuracy on test data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [32]:
print('Accuracy score on Test Data : ', test_data_accuracy)

Accuracy score on Test Data :  0.9741379310344828


# **Overview of Output and Results**
The Credit Card Fraud Detection System was evaluated using Machine Learning, yielding high accuracy on both training and test datasets.

Training Accuracy: 99.14% – Indicates the model has learned patterns effectively from the training data.
Test Accuracy: 97.41% – Demonstrates strong generalization to unseen data, with minimal overfitting.
These results suggest that the model is highly effective in detecting fraudulent transactions while maintaining reliability on new data. However, further evaluation using precision, recall, and F1-score is recommended to assess real-world performance.